doc/release/upcoming_changes/22539.change.rst

```diff
@@ -0,0 +1,19 @@
+``np.r_[]`` and ``np.c_[]`` with certain scalar values
+------------------------------------------------------
+In rare cases, using mainly ``np.r_`` with scalars can lead to different
+results.  The main potential changes are highlighted by the following::
+
+    >>> np.r_[np.arange(5, dtype=np.uint8), -1].dtype
+    int16  # rather than the default integer (int64 or int32)
+    >>> np.r_[np.arange(5, dtype=np.int8), 255]
+    array([  0,   1,   2,   3,   4, 255], dtype=int16)
+
+Where the second example returned::
+
+    array([ 0,  1,  2,  3,  4, -1], dtype=int8)
```

Comment:
IIUC, the current behavior (i.e. 1.23) is:

```python
>>> np.r_[np.arange(5, dtype=np.int8), 255]
array([ 0,  1,  2,  3,  4, -1], dtype=int8)
```

If this PR is merged, the behavior will become:

```python
>>> np.r_[np.arange(5, dtype=np.int8), 255]
array([  0,   1,   2,   3,   4, 255], dtype=int16)
```

But then NEP 50 does away with inspecting the value, so will the behavior change again? And if so, to what?

Reply:
Due to the fact that it is converted to an array until we get there, it would overflow.  But, I could change that to use `np.asarray(..., dtype=dtype)`, in which case we would hit the new `np.int8(255)` is going to fail DeprecationWarning, and it would go to an error.

That seems like the right choice to me, do you agree?